File: src/queries/compare.ts
import { rawQuery } from "@/lib/queries";

export interface ComparisonStats {
  period: "current" | "previous";
  pageviews: number;
  visitors: number;
  visits: number;
  bounces: number;
  totaltime: number;
}

/**
 * Get stats for two periods for comparison.
 */
export async function getComparisonStats(
  websiteId: string,
  currentStart: Date,
  currentEnd: Date,
  previousStart: Date,
  previousEnd: Date
): Promise<ComparisonStats[]> {
  const results = await rawQuery<{
    period: string;
    pageviews: bigint;
    visitors: bigint;
    visits: bigint;
    bounces: bigint;
    totaltime: bigint;
  }>(
    `SELECT
      'current' AS period,
      COALESCE(SUM(t.c), 0)::bigint AS pageviews,
      COUNT(DISTINCT t.session_id)::bigint AS visitors,
      COUNT(DISTINCT t.visit_id)::bigint AS visits,
      COALESCE(SUM(CASE WHEN t.c = 1 THEN 1 ELSE 0 END), 0)::bigint AS bounces,
      COALESCE(SUM(FLOOR(EXTRACT(EPOCH FROM (t.max_time - t.min_time)))), 0)::bigint AS totaltime
    FROM (
      SELECT session_id, visit_id, COUNT(*) AS c,
             MIN(created_at) AS min_time, MAX(created_at) AS max_time
      FROM website_event
      WHERE website_id = {{websiteId::uuid}}
        AND created_at BETWEEN {{currentStart::timestamptz}} AND {{currentEnd::timestamptz}}
        AND event_type = 1
      GROUP BY 1, 2
    ) AS t
    UNION ALL
    SELECT
      'previous' AS period,
      COALESCE(SUM(t.c), 0)::bigint AS pageviews,
      COUNT(DISTINCT t.session_id)::bigint AS visitors,
      COUNT(DISTINCT t.visit_id)::bigint AS visits,
      COALESCE(SUM(CASE WHEN t.c = 1 THEN 1 ELSE 0 END), 0)::bigint AS bounces,
      COALESCE(SUM(FLOOR(EXTRACT(EPOCH FROM (t.max_time - t.min_time)))), 0)::bigint AS totaltime
    FROM (
      SELECT session_id, visit_id, COUNT(*) AS c,
             MIN(created_at) AS min_time, MAX(created_at) AS max_time
      FROM website_event
      WHERE website_id = {{websiteId::uuid}}
        AND created_at BETWEEN {{previousStart::timestamptz}} AND {{previousEnd::timestamptz}}
        AND event_type = 1
      GROUP BY 1, 2
    ) AS t`,
    { websiteId, currentStart, currentEnd, previousStart, previousEnd }
  );

  return results.map((r) => ({
    period: r.period as "current" | "previous",
    pageviews: Number(r.pageviews),
    visitors: Number(r.visitors),
    visits: Number(r.visits),
    bounces: Number(r.bounces),
    totaltime: Number(r.totaltime),
  }));
}

File: core/mantecato_core/queries/compare.py
"""Comparison queries — aggregate pageview comparison between two periods.

Privacy-first: only pageview counts are compared. No visitor/session metrics.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from datetime import datetime

from core.mantecato_core.database import raw_query
from core.mantecato_core.filters import Filter, prepare_filters
from core.mantecato_core.queries.orm_fallbacks import (
    pageview_queryset,
    should_use_orm_fallback,
    stats_dict,
)


def get_comparison_stats(
    website_id: str,
    cur_start: datetime,
    cur_end: datetime,
    prev_start: datetime,
    prev_end: datetime,
    filters: list[Filter] | None = None,
) -> list[dict[str, Any]]:
    """Return pageview counts for two periods for comparison.

    Returns ``[{"period": "current", "pageviews": N}, {"period": "previous", "pageviews": N}]``.
    """
    if should_use_orm_fallback():
        return [
            {"period": "current", **stats_dict(pageview_queryset(website_id, cur_start, cur_end, filters))},
            {"period": "previous", **stats_dict(pageview_queryset(website_id, prev_start, prev_end, filters))},
        ]

    filter_where, filter_params, _ = prepare_filters(filters or [])
    rows = raw_query(
        """SELECT 'current' AS period,
      COUNT(*)::bigint AS pageviews,
      COUNT(*) FILTER (WHERE COALESCE(we.is_bot, false) = false)::bigint AS human_pageviews,
      COUNT(*) FILTER (WHERE COALESCE(we.is_bot, false) = true)::bigint AS bot_pageviews
    FROM website_event we
    WHERE we.website_id = {{websiteId::uuid}}
      AND we.created_at BETWEEN {{curStart::timestamptz}} AND {{curEnd::timestamptz}}
      AND we.event_type = 1
      """ + filter_where + """
    UNION ALL
    SELECT 'previous' AS period,
      COUNT(*)::bigint AS pageviews,
      COUNT(*) FILTER (WHERE COALESCE(we.is_bot, false) = false)::bigint AS human_pageviews,
      COUNT(*) FILTER (WHERE COALESCE(we.is_bot, false) = true)::bigint AS bot_pageviews
    FROM website_event we
    WHERE we.website_id = {{websiteId::uuid}}
      AND we.created_at BETWEEN {{prevStart::timestamptz}} AND {{prevEnd::timestamptz}}
      AND we.event_type = 1
      """ + filter_where,
        {
            "websiteId": website_id,
            "curStart": cur_start,
            "curEnd": cur_end,
            "prevStart": prev_start,
            "prevEnd": prev_end,
            **filter_params,
        },
    )
    return [
        {
            "period": row["period"],
            "pageviews": int(row.get("pageviews") or 0),
            "human_pageviews": int(row.get("human_pageviews") or 0),
            "bot_pageviews": int(row.get("bot_pageviews") or 0),
        }
        for row in rows
    ]
